<a href="https://colab.research.google.com/github/digitechit07/Python-Tutorial-with-Excercise/blob/main/Python_Math_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Number-theoretic and representation functions**
math.ceil(x)
Return the ceiling of x, the smallest integer greater than or equal to x. If x is not a float, delegates to x.__ceil__(), which should return an Integral value.

math.copysign(x, y)
Return a float with the magnitude (absolute value) of x but the sign of y. On platforms that support signed zeros, copysign(1.0, -0.0) returns -1.0.

math.fabs(x)
Return the absolute value of x.

math.factorial(x)
Return x factorial. Raises ValueError if x is not integral or is negative.

math.floor(x)
Return the floor of x, the largest integer less than or equal to x. If x is not a float, delegates to x.__floor__(), which should return an Integral value.

math.fmod(x, y)

In [ ]:
import math #imports math module

import math as m
print(m.sqrt(25))

from math import sqrt #imports a method from math module
print(sqrt(25))

from math import sqrt as s
print(s(25))

from math import *
print(sqrt(25))


import math
math.sqrt(4)


import math as m
a=int(input("Enter the no"))
print(m.sqrt(a))

from math import sqrt
print(sqrt(25))

from math import sqrt as s
print(s(25))

from math import *
print(sqrt(25))

a = 5
b = 7
c = a + b
print(c)

a = 16
b = 8
c = 3
print(a - b + c)

a = 11
b = 5
c = a/b
print(c)

a = 11
b = 5
c = 10
print(a/b*c)

a = 11
b = 5
print(a%b)

a = [1,2,3,4,5,6]
for x in a:
   if x%2==0:
      print(x)

a = 2
b = 5
print(a**b)
a = 2
b = 5
c = 15
d = 3
print(a**b+c/d)

def test(n):
    return (n * (n + 1) * (n + 2)) / 6

n = 1
print("\nOriginal Number:",n)
print("Tetrahedral number:",test(n))
n = 2
print("\nOriginal Number:",n)
print("Tetrahedral number:",test(n))
n = 6
print("\nOriginal Number:",n)
print("Tetrahedral number:",test(n))


def sum_of_powers(end, power = 2, start = 1):
  return sum([(i) ** power for i in range(start, end + 1)])

print(sum_of_powers(12))
print(sum_of_powers(12, 3))
print(sum_of_powers(12, 5, 7))


def hamming_distance(x, y):
  return bin(x ^ y).count('1')
x = 2
y = 3
print("Hamming distance between",x,"and",y,"is",hamming_distance(x, y))
x = 43
y = 87
print("\nHamming distance between",x,"and",y,"is",hamming_distance(x, y))

def clamp_number(num, a, b):
  return max(min(num, max(a, b)), min(a, b))

print(clamp_number(2, 4, 7))
print(clamp_number(1, -1, -4))
print(clamp_number(22, -10, 10))
print(clamp_number(20, 1, 10))
print(clamp_number(5, 1, 20))
print(clamp_number(1, 10, 20))

def is_disarium(num):
    temp = 0
    for i in range(len(str(num))):
        temp += int(str(num)[i]) ** (i + 1)
    return temp == num

num = 25
print("\nIs",num,"is Disarium number?",is_disarium(num))
num = 89
print("\nIs",num,"is Disarium number?",is_disarium(num))
num = 75
print("\nIs",num,"is Disarium number?",is_disarium(num))
num = 125
print("\nIs",num,"is Disarium number?",is_disarium(num))
num = 518
print("\nIs",num,"is Disarium number?",is_disarium(num))


def test(n):
    if n < 0: return False
    if n == 0: return True
    return len(set(str(n))) == 1
n = 0
print("Original number:", n)
print("Check the said number is a repdigit number or not!")
print(test(n))
n = -1111
print("Original number:", n)
print("Check the said number is a repdigit number or not!")
print(test(n))
n = 1
print("Original number:", n)
print("Check the said number is a repdigit number or not!")
print(test(n))
n = 9999999
print("Original number:", n)
print("Check the said number is a repdigit number or not!")
print(test(n))


def test(n):
    if (n>0):
        a = 0
        b = n
        while b > 0:
            a = a +  b % 10
            b = b // 10
        return not n % a

n = 666
print("Original number:", n)
print("Check the said number is a Harshad number or not!")
print(test(n))
n = 11
print("\nOriginal number:", n)
print("Check the said number is a Harshad number or not!")
print(test(n))
n = -144
print("\nOriginal number:", n)
print("Check the said number is a Harshad number or not!")
print(test(n))
n = 200
print("\nOriginal number:", n)
print("Check the said number is a Harshad number or not!")
print(test(n))


def test(n):
	n += 1
	if len(set(str(n))) == len(str(n)):
		return n
	#return test(n)
n = 12345
print("Original number:", n)
print("Next number containing only distinct digits!")
print(test(n))
n = 99999
print("Original number:", n)
print("Next number containing only distinct digits!")
print(test(n))
n = 1
print("Original number:", n)
print("Next number containing only distinct digits!")
print(test(n))
n = 100
print("Original number:", n)
print("Next number containing only distinct digits!")
print(test(n))




5.0
5.0
5.0
5.0
Enter the no3
1.7320508075688772
5.0
5.0
5.0
12
11
2.2
22.0
1
2
4
6
32
37.0

Original Number: 1
Tetrahedral number: 1.0

Original Number: 2
Tetrahedral number: 4.0

Original Number: 6
Tetrahedral number: 56.0
650
6084
618507
Hamming distance between 2 and 3 is 1

Hamming distance between 43 and 87 is 5
4
-1
10
10
5
10

Is 25 is Disarium number? False

Is 89 is Disarium number? True

Is 75 is Disarium number? False

Is 125 is Disarium number? False

Is 518 is Disarium number? True
Original number: 0
Check the said number is a repdigit number or not!
True
Original number: -1111
Check the said number is a repdigit number or not!
False
Original number: 1
Check the said number is a repdigit number or not!
True
Original number: 9999999
Check the said number is a repdigit number or not!
True
Original number: 666
Check the said number is a Harshad number or not!
True

Original number: 11
Check the said number is a Harshad number or not!
False

Original number: -144
Check the sa

# **Power and logarithmic functions**
math.exp(x)
Return e**x.

math.expm1(x)
Return e**x - 1. For small floats x, the subtraction in exp(x) - 1 can result in a significant loss of precision; the expm1() function provides a way to compute this quantity to full precision:

>>>
>>> from math import exp, expm1
>>> exp(1e-5) - 1  # gives result accurate to 11 places
1.0000050000069649e-05
>>> expm1(1e-5)    # result accurate to full precision
1.0000050000166668e-05
Nouveau dans la version 3.2.

math.log(x[, base])
With one argument, return the natural logarithm of x (to base e).

With two arguments, return the logarithm of x to the given base, calculated as log(x)/log(base).

math.log1p(x)
Return the natural logarithm of 1+x (base e). The result is calculated in a way which is accurate for x near zero.

math.log2(x)

In [ ]:
def test(n):
    return all(abs(int(x) - int(y)) == 2 for x, y in zip(str(n), str(n)[1:]))

n = 666
print("Original number:", n)
print("Is absolute difference between two consecutive digits of the said number is 2 or not?")
print(test(n))
n = 3579
print("\nOriginal number:", n)
print("Is absolute difference between two consecutive digits of the said number is 2 or not?")
print(test(n))
n = 2468
print("\nOriginal number:", n)
print("Is absolute difference between two consecutive digits of the said number is 2 or not?")
print(test(n))
n = 20420
print("\nOriginal number:", n)
print("Is absolute difference between two consecutive digits of the said number is 2 or not?")
print(test(n))


def test(n):
    temp = []
    for e in str(n):
        temp.append(e)
        max_num = "".join(sorted(temp)[::-1])
        min_num = "".join(sorted(temp))
    return int(max_num), int(min_num)

n = 1254
print("Original number:", n)
print("Rearrange the digits of the said number to get Maximum and Minimum numbers:")
print("Maximum and Minimum Numbers:",test(n))
n = 6
print("\nOriginal number:", n)
print("Rearrange the digits of the said number to get Maximum and Minimum numbers:")
print("Maximum and Minimum Numbers:",test(n))
n = 1000
print("\nOriginal number:", n)
print("Rearrange the digits of the said number to get Maximum and Minimum numbers:")
print("Maximum and Minimum Numbers:",test(n))


def test(nums):
    if len(nums) > 0:
        return sum(list(filter(lambda x: (x > 1 and all(x % y != 0 for y in range(2, x))), nums)))
    return "Empty list!"

nums = [1, 3, 4, 7, 9]
print("Original list:")
print(nums)
print("Sum of all prime numbers in the said list of numbers:")
print(test(nums))
nums = []
print("\nOriginal list:")
print(nums)
print("Sum of all prime numbers in the said list of numbers:")
print(test(nums))
nums = [11, 37, 444]
print("\nOriginal list:")
print(nums)
print("Sum of all prime numbers in the said list of numbers:")
print(test(nums))

# Addition
result = 2 + 3
print(result)

# Subtraction
result = 5 - 2
print(result)

# Asks the user to input a temperature value in Fahrenheit
fahrenheit = int(input("Fahrenheit Temperature: "))

# Formula to calculate temperature in F from C
celsius = (fahrenheit - 32) * (5/9)

# Rounding the variable celsius to the nearest integer
celsius_rounded = round(celsius)

# Prints original Fahrenheit temperature and calculated Celsius temperature
print(str(fahrenheit) + " degrees Fahrenheit = " + str(celsius_rounded) + " degrees Celsius")



first = 15 % 6
second = 15 % 5
third = 15 % 4
fourth = 15 % 3
fifth = 15 % 2

print(first)
print(second)
print(third)
print(fourth)
print(fifth)

x = 2
y = 3
z = 5

# Perform one calculation on itself
x -= 2 # Same as x = x - 2
y *= 3 # Same as y = y * 3
z /= 2 # Same as z = z / 2

print("x: " + str(x))
print("y: " + str(y))
print("z: " + str(z))

# ---------------------------------------------- #

x = 2
y = 3
z = 4

# Same as above, but using the variables for the calculation

x -= y # Same as x = x - y
y *= z # Same as y = y * z


print("")
print("x: " + str(x))
print("y: " + str(y))

import math
# ** operator means to raise power, i.e. ab for a ** b
res = math.e ** 3 + 5
print(res)


import math
radius = 4.5
area = math.pi * radius * radius
print(area)


import math
# ** operator means to raise power, i.e. ab for a ** b
res = math.tau ** 3 + 5
print(res)


import math
print(math.inf)


import math
print(math.nan)


import math
num = math.ceil(2.6)
print(num)


import math
num = math.floor(2.6)
print(num)


import math
val = math.floor(-2.7)
print(val)


import math
num = math.factorial(5)
print(num)

fact = 1 # initialising fact=1
n = 1000000 # the number for which factorial has to be found
for i in range(1,n+1):  # loop will run n times
   fact*=i


import math
num = math.gcd(5,25)
print(num)


import math
num = math.fabs(-3.4)
print(num)


import math
num = math.exp(2.4)
print(num)


import math
num = math.pow(2,4)
print(num)


import math
num = math.log(16,2)
print(num)


import math
res = math.log(math.e)
print(res)


import math
num = math.sqrt(16)
print(num)






Original number: 666
Is absolute difference between two consecutive digits of the said number is 2 or not?
False

Original number: 3579
Is absolute difference between two consecutive digits of the said number is 2 or not?
True

Original number: 2468
Is absolute difference between two consecutive digits of the said number is 2 or not?
True

Original number: 20420
Is absolute difference between two consecutive digits of the said number is 2 or not?
False
Original number: 1254
Rearrange the digits of the said number to get Maximum and Minimum numbers:
Maximum and Minimum Numbers: (5421, 1245)

Original number: 6
Rearrange the digits of the said number to get Maximum and Minimum numbers:
Maximum and Minimum Numbers: (6, 6)

Original number: 1000
Rearrange the digits of the said number to get Maximum and Minimum numbers:
Maximum and Minimum Numbers: (1000, 1)
Original list:
[1, 3, 4, 7, 9]
Sum of all prime numbers in the said list of numbers:
10

Original list:
[]
Sum of all prime numbers i

# **Hyperbolic functions**
Hyperbolic functions are analogs of trigonometric functions that are based on hyperbolas instead of circles.

math.acosh(x)
Return the inverse hyperbolic cosine of x.

math.asinh(x)
Return the inverse hyperbolic sine of x.

math.atanh(x)
Return the inverse hyperbolic tangent of x.

math.cosh(x)
Return the hyperbolic cosine of x.

math.sinh(x)
Return the hyperbolic sine of x.

math.tanh(x)
Return the hyperbolic tangent of x.

In [2]:
import math
num = math.sin(1.46)
print(num)


import math
num = math.cos(1.46)
print(num)


import math
num = math.tan(1.46)
print(num)


import math
num = math.degrees(math.pi)
print(num)


import math
num = math.radians(180.0)
print(num)


import math
val = math.gamma(5.7)
print(val)


import math
res = math.isinf(6)
print(res)


import math
res = math.isinf(math.inf)
print(res)


import math
res = math.isnan(6)
print(res)


import math
res = math.isfinite(6)
print(res)


import math
res = math.isfinite(math.inf)
print(res)


# Import math Library
import math

# Print the value of pi
print (math.pi)

import math

angle_degree = 60
angle_radian = math.radians(angle_degree)

print('The input angle: ', angle_radian)
print('sin(x): ', math.sin(angle_radian))
print('cos(x): ', math.cos(angle_radian))
print('tan(x): ', math.tan(angle_radian))

import math
print('The value of 2^2: ' + str(math.pow(2, 2)))
print('Square root of 121: ' + str(math.sqrt(121)))
print('The value of 8^e: ' + str(math.exp(8)))
print('The value of Log(625) with base 5: ' + str(math.log(625, 5)))
print('The value of Log(444) with base 2: ' + str(math.log2(444)))
print('The value of Log(1000) with base 10: ' + str(math.log10(1000)))

import math

input = 12.35
print('The Floor value of the given input: ' + str(math.floor(input)))

print('The Ceil value of the given input: ' + str(math.ceil(input)))

a = 20
b = -10

print('The value of a after copying the sign from b: ' + str(math.copysign(a, b)))

s1 = -25
s2 = 25

print('Absolute value of s1 and s2: ' + str(math.fabs(s1)) + ', ' + str(math.fabs(s2)))

my_input = [0,1,2,3,4,5,6,7,8,9,-1]
print('Sum of the elements of the list: ' + str(math.fsum(my_input)))

p = float('nan')
if math.isnan(p):
    print('It is not a number')
q = float('inf')
y = 10
if math.isinf(q):
    print('It is Infinity')
print(math.isfinite(q)) #q is not a finite number
print(math.isfinite(y)) #y is a finite number

import math

print('Value of pi: ', math.pi)

radius = 2

print('Area of Circle: ', math.pi * (radius ** 2))

import math
print(f"π = {math.pi}")
print(f"e = {math.e}")
print(f"τ = {math.tau}")
print(f"Infinity: {math.inf}")

import math
# Ceiling and floor operations
print(math.ceil(4.3))    # Output: 5
print(math.floor(4.7))   # Output: 4
print(math.trunc(-4.7))  # Output: -4
# GCD and LCM
print(math.gcd(48, 18))  # Output: 6
print(math.lcm(12, 18))  # Output: 36

import math
# Factorial
print(math.factorial(5))  # Output: 120
# Permutations and combinations
print(math.perm(5, 2))    # Output: 20
print(math.comb(5, 2))    # Output: 10

import math
# Exponential functions
print(math.exp(2))        # e^2
print(math.pow(2, 3))     # 2^3 = 8
# Logarithmic functions
#print(math.log(100, 10))  # log₁₀(100) = 2
#print(math.log2(8))       # log₂(8) = 3
#print(math.ln(math.e))    # ln(e) = 1

import math
# Convert degrees to radians
angle_degrees = 45
angle_radians = math.radians(angle_degrees)
# Trigonometric calculations
print(f"sin(45°) = {math.sin(angle_radians):.4f}")
print(f"cos(45°) = {math.cos(angle_radians):.4f}")
print(f"tan(45°) = {math.tan(angle_radians):.4f}")
# Convert back to degrees
result = math.degrees(math.asin(0.5))
print(f"arcsin(0.5) = {result}°")

import math
# Accurate sum for floating-point numbers
numbers = [0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1]
print(sum(numbers))        # May have floating-point errors
print(math.fsum(numbers))  # More accurate result
# Product of numbers
values = [2, 3, 4, 5]
print(math.prod(values))   # Output: 120

import math
# Testing special values
print(math.isfinite(5.0))      # True
print(math.isinf(math.inf))    # True
print(math.isnan(math.nan))    # True
# Comparing floating-point numbers
a = 0.1 + 0.1 + 0.1
b = 0.3
print(a == b)                  # False (floating-point precision)
print(math.isclose(a, b))      # True (within tolerance)

import math
# 2D distance
print(math.hypot(3, 4))  # √(3² + 4²) = 5
# 3D distance
print(math.hypot(1, 2, 3))  # √(1² + 2² + 3²)
# Distance between points
point1 = (1, 2)
point2 = (4, 6)
print(math.dist(point1, point2))  # Distance between points

import math
def compound_interest(principal, rate, time, n=1):
    """Calculate compound interest"""
    amount = principal * math.pow((1 + rate/n), n * time)
    return amount
# Example: $1000 at 5% annual interest for 3 years
principal = 1000
rate = 0.05
time = 3
result = compound_interest(principal, rate, time)
print(f"Amount after {time} years: ${result:.2f}")

import math
def rotate_point(x, y, angle_degrees):
    """Rotate a point around the origin"""
    angle_rad = math.radians(angle_degrees)
    cos_a = math.cos(angle_rad)
    sin_a = math.sin(angle_rad)

    new_x = x * cos_a - y * sin_a
    new_y = x * sin_a + y * cos_a

    return new_x, new_y
# Rotate point (1, 0) by 90 degrees
new_x, new_y = rotate_point(1, 0, 90)
print(f"Rotated point: ({new_x:.6f}, {new_y:.6f})")

import math
def standard_deviation(data):
    """Calculate standard deviation using math functions"""
    n = len(data)
    mean = math.fsum(data) / n
    variance = math.fsum((x - mean) ** 2 for x in data) / n
    return math.sqrt(variance)
# Example dataset
data = [2, 4, 4, 4, 5, 5, 7, 9]
std_dev = standard_deviation(data)
print(f"Standard deviation: {std_dev:.4f}")

0.9938683634116449
0.11056977982006959
8.988607601724167
180.0
3.141592653589793
72.5276345202229
False
True
False
True
False
3.141592653589793
The input angle:  1.0471975511965976
sin(x):  0.8660254037844386
cos(x):  0.5000000000000001
tan(x):  1.7320508075688767
The value of 2^2: 4.0
Square root of 121: 11.0
The value of 8^e: 2980.9579870417283
The value of Log(625) with base 5: 4.0
The value of Log(444) with base 2: 8.794415866350105
The value of Log(1000) with base 10: 3.0
The Floor value of the given input: 12
The Ceil value of the given input: 13
The value of a after copying the sign from b: -20.0
Absolute value of s1 and s2: 25.0, 25.0
Sum of the elements of the list: 44.0
It is not a number
It is Infinity
False
True
Value of pi:  3.141592653589793
Area of Circle:  12.566370614359172
π = 3.141592653589793
e = 2.718281828459045
τ = 6.283185307179586
Infinity: inf
5
4
-4
6
36
120
20
10
7.38905609893065
8.0
sin(45°) = 0.7071
cos(45°) = 0.7071
tan(45°) = 1.0000
arcsin(0.5) = 30.0000

# **Constants**
math.pi
The mathematical constant π = 3.141592..., to available precision.

math.e
The mathematical constant e = 2.718281..., to available precision.

math.tau
The mathematical constant τ = 6.283185..., to available precision. Tau is a circle constant equal to 2π, the ratio of a circle’s circumference to its radius. To learn more about Tau, check out Vi Hart’s video Pi is (still) Wrong, and start celebrating Tau day by eating twice as much pie!

Nouveau dans la version 3.6.

math.inf
A floating-point positive infinity. (For negative infinity, use -math.inf.) Equivalent to the output of float('inf').

Nouveau dans la version 3.5.

math.nan
A floating-point “not a number” (NaN) value. Equivalent to the output of float('nan').